src/consensus.rs

```diff
@@ -244,18 +244,21 @@ impl<T: Proposition> Consensus<T> {
         })
     }
 
-    pub fn cast_vote(&mut self, signed_vote: &SignedVote<T>) -> Result<SignedVote<T>> {
+    pub fn cast_vote(&mut self, signed_vote: SignedVote<T>) -> Result<SignedVote<T>> {
+        info!("[{}] casting vote {:?}", self.id(), signed_vote);
         match self.handle_signed_vote(signed_vote.clone())? {
-            VoteResponse::WaitingForMoreVotes => Ok(signed_vote.clone()),
+            VoteResponse::WaitingForMoreVotes => Ok(signed_vote),
             VoteResponse::Broadcast(vote) => Ok(vote),
         }
     }
 
     pub fn log_signed_vote(&mut self, signed_vote: &SignedVote<T>) {
         for vote in signed_vote.unpack_votes() {
-            let existing_vote = self.votes.entry(vote.voter).or_insert_with(|| vote.clone());
-            if vote.supersedes(existing_vote) {
-                *existing_vote = vote.clone()
+            if self.seen_votes_cache.insert(vote.sig.clone()) {
```

Comment:
`BTreeSet::insert(m)` returns `true` when `m` was not a member of the set, and `false` if it was already a member